packages/compass-editor/src/json-editor.tsx

```diff
@@ -838,11 +900,45 @@ const BaseEditor = React.forwardRef<EditorRef, EditorProps>(function BaseEditor(
         readOnlyExtension,
         EditorView.updateListener.of((update) => {
           updateEditorContentHeight();
+          const editorText = editor.state.sliceDoc() ?? '';
+
+          if (update.focusChanged && editorViewRef.current?.hasFocus) {
+            const editorView = editorViewRef.current;
+
+            onFocusRef.current?.({ editorView, setContent }, editorText);
+          }
+
           if (update.docChanged) {
-            const text = editor.state.sliceDoc() ?? '';
-            onChangeTextRef.current?.(text, update);
+            onChangeTextRef.current?.(editorText, update);
           }
         }),
+        EditorView.domEventHandlers({
+          paste(event, view) {
+            const initialState = view.state.sliceDoc() ?? '';
+            const clipboardContents =
+              event.clipboardData?.getData('text/plain') ?? '';
+            const selection = view.state.selection.main;
+            const isReplace =
```

Comment:
But you have this dependency: query bar field editor needs to apply some custom logic to modify user input in some very special corner cases, why editor should be the one handling this? I think I'm starting to understand a bit: are you trying to bake all this custom logic in the general purpose editor component just to avoid using codemirror ref in the parent? I strongly believe it's not the right way of approaching this and you are anyway exposing the internals here, just with more steps and breaking existing interface established in this component that follow React patterns. Imperative handles are as much a pattern in React as component state or props, its use-cases are rare, yes, but we should not avoid it if that's the right tool to use.